server/persistence/dynamo/utils.go

```diff
@@ -0,0 +1,106 @@
+package dynamo
+
+import (
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+)
+
+const (
+	dbName       = `cribbage`
+	partitionKey = `DDBid`
+	sortKey      = `spec`
+)
+
+func getSortKeyPrefix(service interface{}) string {
+	// all of these need to be different, because they are the
+	// start of the sort key. we are partitioning our dynamo table usage
+	// such that each service has the same prefix:#
+	switch service.(type) {
+	case *gameService:
+		return `game`
+	case *interactionService:
+		return `interaction`
+	case *playerService:
+		return `player`
+	}
+
+	return `garbage`
+}
+
+type condExprType uint8
+
+const (
+	none      condExprType = 0
+	equalsID  condExprType = 1
+	hasPrefix condExprType = 2
+	notExists condExprType = 3
+)
+
+func getConditionExpression(
+	pkType condExprType,
+	pk string,
+	skType condExprType,
+	sk string,
+) *string {
```

Comment:
This function does some disparate things depending on the inputs; an empty `pk` or `sk` is not required with `notExists` as compared to e.g. `equalsID` where you do need to provide those. This makes it a little tough to mentally parse calling sites. I wonder if it'd benefit from taking a struct as a single input arg instead?

Reply:
:bulb: I don't know if you'll like this idea since it's passing a lot of behavior around and it's kind of a lot of boilerplate -- but I kind of like this pattern as a way to provide composable options to a function
```go
func getConditionExpression(opts ...conditionExpressionOption) *string {
    opts := conditionExpressionOptions{}
    for _, opt := range opts {
        opts = opt(opts)
    }
    // do whatever with your options
}

type conditionExpressionOptions {
    pkType string
    pk string
    skType string
    sk string
}

type conditionExpressionOption func(opts conditionExpressionOptions) conditionExpressionOptions

func pkNotExists(opts conditionExpressionOptions) conditionExpressionOptions {
    opts.pkType = notExists
    return opts
}

func skHasPrefix(pref string) conditionExpressionOption {
    return func(opts conditionExpressionOptions) conditionExpressionOptions {
        opts.skType = hasPrefix
        opts.sk = pref
        return opts
    }
}
// etc.

// calling sites
getConditionExpression(pkNotExists)
getConditionExpression(pkEquals(`def`), skHasPrefix(`abc`))
```